Adding user groups via release coordinator page

lkbhitesh07 opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of N/A.
2. This PR does the following: Introduces a way to add user-groups via the release-coordinator page. We can assign usernames to certain user-groups.
3. (For bug-fixing PRs only) The original bug occurred because: N/A

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [ ] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [ ] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

## Proof that changes are correct

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

Following video shows the updates made to user-groups. We are adding new user-group, adding users to user-groups, resetting the changes if not needed. Deleting the user group and creating a new one with the same name.


https://github.com/user-attachments/assets/573b68f1-a335-4903-a58b-443e5352fe79


## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

lkbhitesh07 committed:
Changes to add user groups via admin page

lkbhitesh07 committed:
Fixed the failing tests

lkbhitesh07 committed:
Updates to fix the tests

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Fixed multiple tests

lkbhitesh07 committed:
Frontend test fixed

lkbhitesh07 commented:
Hey @vojtechjelinek @Nik-09, PTAL whenever you get some time.
Thanks!

vojtechjelinek reviewed (approved):
LGTM for codeowner files.

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

Nik-09 reviewed (changes requested):
Hi @lkbhitesh07 I looked into the UI, what happens if I type the username for which any user does not exist, so can you please provide video proof for that?
Thank you

lkbhitesh07 commented:
Hi @Nik-09, Thanks for the review. Actually, we are not planning to get the list of complete usernames and then verify against them, it would be a time consuming task as we have such a large number of entries for the users. Also, whenever we will load the screen the users will be fetched from the storage. The plan is that the admin will simply be adding the usernames to the user-groups, if the username exists, the feature-flag will be turned on for that user. Let me also confirm this with @seanlip.
Hi @seanlip, please let me know if the above one is the correct behavior.
Thanks

StephenYu2018 commented on core/domain/user_services_test.py:
```diff
@@ -1775,6 +1775,44 @@ def test_mark_user_for_deletion_marks_user_settings_as_deleted(
+            'USER_GROUP_1': ['user_1_id', 'user_2_id', 'user_3_id'],
+            'USER_GROUP_2': ['user_1_id', 'user_2_id']
+        })
+        user_models.UserGroupModel.delete_multi(
+            [
+                user_models.UserGroupModel.get('USER_GROUP_1'),
+                user_models.UserGroupModel.get('USER_GROUP_2')
+            ]
+        )
```
Are we deleting the models created for this test as clean up for the next test? If so, then why do other tests that create `UserGroupModel` not have the same thing?

lkbhitesh07 replied:
Yes, I have updated it now. Thanks for pointing it out.

StephenYu2018 commented on core/domain/user_services_test.py:
```diff
@@ -1775,6 +1775,44 @@ def test_mark_user_for_deletion_marks_user_settings_as_deleted(
         assert user_settings is not None
         self.assertTrue(user_settings.deleted)
 
+    def test_get_all_user_groups_present_in_storage_layer(self) -> None:
```
Does any part of the codebase who calls `user_services.get_user_group_models_dict()` need to know that `user_services` gets the user group data from persistence or through the network? If not, then I think we should remove the `_present_in_storage_layer` part of this test name.

lkbhitesh07 replied:
Updated

StephenYu2018 commented on core/domain/user_services.py:
```diff
@@ -495,6 +495,53 @@ def get_user_settings_by_auth_id(
         return None
 
 
+def get_user_group_models_dict() -> Dict[str, List[str]]:
```
Rename to `get_all_user_groups`. The word `models` is redundant given that the code is trying to model/represent a user group. The word `dict` already appears in the type signature, so it's also redundant.

lkbhitesh07 replied:
Updated

StephenYu2018 commented on core/domain/user_services.py:
```diff
@@ -495,6 +495,53 @@ def get_user_settings_by_auth_id(
+            user_group_model.update_timestamps()
+            user_group_model.put()
+    # Delete UserGroup models.
+    all_exisiting_user_groups: List[
```
Typo: `exisiting` should be `existing`

lkbhitesh07 replied:
Updated

StephenYu2018 commented on core/domain/user_services.py:
```diff
@@ -495,6 +495,53 @@ def get_user_settings_by_auth_id(
+            user_models.UserGroupModel.get_all()
+        )
+    user_models_to_delete = []
+    for exising_user_group in all_exisiting_user_groups:
```
Typo: `exising` should be `existing`

lkbhitesh07 replied:
Updated

StephenYu2018 commented on core/templates/pages/admin-page/misc-tab/admin-misc-tab.component.ts:
```diff
@@ -16,18 +16,34 @@
   templateUrl: './admin-misc-tab.component.html',
 })
 export class AdminMiscTabComponent {
+  @ViewChild('userInput') userInput!: ElementRef<HTMLInputElement>;
```
I don't think the ref should be called `userInput` on its own because it's too vague. From this PR, this ref seems to only be introduced for managing user groups, but the name `userInput` could imply any user input on the admin misc tab (like revoking super admin privileges, and looking up exploration interactions). Maybe try `userInputToAddUserToGroup`

lkbhitesh07 replied:
Updated

StephenYu2018 commented on core/templates/pages/admin-page/misc-tab/admin-misc-tab.component.ts:
```diff
@@ -60,13 +76,138 @@ export class AdminMiscTabComponent {
+    return userGroups;
+  }
+
+  addUser(event: {value: string}): void {
```
Rename method name to `addUserToSelectedGroup`. Recall that the admin misc tab contains different functionalities unrelated to adding users to user groups.

lkbhitesh07 replied:
Updated

StephenYu2018 commented on core/templates/pages/admin-page/misc-tab/admin-misc-tab.component.ts:
```diff
@@ -60,13 +76,138 @@ export class AdminMiscTabComponent {
+    this.userInput.nativeElement.value = '';
+  }
+
+  removeUser(username: string): void {
```
Rename method name to `removeUserFromSelectedGroup`.

lkbhitesh07 replied:
Updated

StephenYu2018 commented on core/templates/pages/admin-page/misc-tab/admin-misc-tab.component.ts:
```diff
@@ -60,13 +76,138 @@ export class AdminMiscTabComponent {
+          this.adminTaskManagerService.finishTask();
+        },
+        errorResponse => {
+          this.setStatusMessage.emit('Server error: ' + errorResponse);
```
Use string interpolation instead of concatenation.

lkbhitesh07 replied:
Updated

StephenYu2018 commented on core/templates/pages/admin-page/misc-tab/admin-misc-tab.component.ts:
```diff
@@ -60,13 +76,138 @@ export class AdminMiscTabComponent {
+  loadingMessage: string = '';
+  selectedUserGroup: string = '';
+  newUserGroupName: string = '';
+  removable: boolean = true;
+  selectable: boolean = true;
```
What is `removable` and `selectable`? Do both of these refer to user groups or the users in these groups? Maybe rename it to `userGroupIsRemovable` or `userIsRemovableFromUserGroup`. Ditto for `selectable`.

lkbhitesh07 replied:
Well come to think of it we would not be requiring these variables because as of now all the user-groups will be selectable and removable, hence I have removed them. Thanks for pointing it out!

StephenYu2018 reviewed (changes requested):
Several concerns.

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Addressed the comments

lkbhitesh07 commented:
> Several concerns.

Thanks @StephenYu2018 for the review, I have updated with the changes. PTAL!

seanlip reviewed (changes requested):
> Hi @Nik-09, Thanks for the review. Actually, we are not planning to get the list of complete usernames and then verify against them, it would be a time consuming task as we have such a large number of entries for the users. Also, whenever we will load the screen the users will be fetched from the storage. The plan is that the admin will simply be adding the usernames to the user-groups, if the username exists, the feature-flag will be turned on for that user. Let me also confirm this with @seanlip. Hi @seanlip, please let me know if the above one is the correct behavior. Thanks

Hi @lkbhitesh07 @Nik-09, some thoughts:

- I think it would be good to validate the username as soon as it is added, to help guard against typos. You can do an async request to the server for that on-add and disable the submit button if you detect a problem.
- Generally speaking I don't understand the UI. Why is there an empty dropdown at the outset? What does the cross do, what does the plus do? I think you should have a UI that's a bit like the classroom admin page, where you show a list of the user groups and who is in them, and the user can add a new item to that list or delete items from that list or edit an item in the list. The current UI is not intuitive and quite prone to errors IMO.

Please let me know if you'd like clarification/help with the UI side of things (but as mentioned above I think the classroom-admin page is a good model for this).

Thanks!

Nik-09 reviewed (commented):
Thank you @seanlip for sharing the feedback.
Hi @lkbhitesh07, I am de-assigning myself from the PR; please re-assign once this is ready for review. Thank you

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Updates to the UI

lkbhitesh07 committed:
some updates to the warnings

lkbhitesh07 commented:
Hey @Nik-09 @seanlip, thanks for the review. I have updated the UI, please find the video attached in the PR description.

Hey @StephenYu2018, Just a ping for the review.
Thanks!!

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

StephenYu2018 commented on core/domain/user_services.py:
```diff
@@ -495,6 +495,53 @@ def get_user_settings_by_auth_id(
+    name and value as list of users associated to the user-group.
+
+    Returns:
+        user_group_models_dict: Dict[str, List[str]]. All user groups dictionary
```
Remove `user_group_models_dict: ` in function docstring.

StephenYu2018 replied:
@lkbhitesh07 Btw. This isn't actually outdated yet.

StephenYu2018 reviewed (changes requested):
One more concern.

Nik-09 commented on core/templates/pages/admin-page/misc-tab/admin-misc-tab.component.ts:
```diff
@@ -60,13 +76,154 @@ export class AdminMiscTabComponent {
+    private loaderService: LoaderService
   ) {}
 
+  async fetchUserGroupData(): Promise<void> {
```
```suggestion
  async fetchUserGroupDataAsync(): Promise<void> {
```

Nik-09 reviewed (approved):
Thank you @lkbhitesh07 LGTM. 
One optional suggestion.

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

seanlip commented on core/templates/pages/admin-page/misc-tab/admin-misc-tab.component.ts:
```diff
@@ -60,13 +76,154 @@ export class AdminMiscTabComponent {
+      return;
+    }
+    if (!this.allUsersUsernames.includes(value)) {
+      this.userInUserGroupValidationError = `The user with username '${value}' does not exists.`;
```
exists --> exist

seanlip reviewed (changes requested):
Hi @lkbhitesh07, thanks! Some comments:

- Shouldn't this functionality be on the RC page rather than the admin page, since it's being used to manage groups used for feature rollouts? If it's on /admin then only the super-admin has access to it.
- I am not sure about the "Save/Reset" functionality at the bottom. Generally the behaviour should be: it shows in view-only mode. Then you can click something to edit a group. At the bottom of that edit view, there are "Save" and "Cancel" buttons; the backend save happens only when the Save button is clicked, and changes are discarded when the Cancel button is clicked. There should also be ways to add a new group and delete an existing group, and both those actions can happen instantaneously on the backend once performed in the frontend.
- How would a user change the name of the user group?

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Revamped the UI | Added frontend domain model | Updated backend functions accordingly

lkbhitesh07 committed:
Updates to resolve the errors

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Migrated to release-coordinator page

lkbhitesh07 committed:
fixed tests

lkbhitesh07 commented:
Hi @seanlip, Thanks for the review. I have completed the changes as requested. Please have a look.

Hi @DubeySandeep @U8NWXD @StephenYu2018  Please have a look.

Please note that currently I am facing one of the test failures which I am trying to figure out, pasting the ss for the reference. I have introduced one modal for deleting the user-group and due to that I am getting the "scss file not present" error, as far as I have note we do not really have scss file for modals. Please let me know if you have some info to avoid the error.
Failing CI test link - [link](https://github.com/oppia/oppia/actions/runs/10136933232/job/28026571162?pr=20429)
Thanks
![Screenshot from 2024-07-29 08-39-36](https://github.com/user-attachments/assets/7aa1fb60-dde5-4c3a-a197-429d37c643ee)

U8NWXD commented:
@vojtechjelinek has already approved for dev workflow reviewers, so I'm removing my assignment. Feel free to re-assign me if there's anything you want my opinion on!

DubeySandeep reviewed (approved):
LGTM for the changes in `.github/CODEOWNERS`

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Fixed the build error by removing the reference for the scss file from component

lkbhitesh07 commented:
> Hi @seanlip, Thanks for the review. I have completed the changes as requested. Please have a look.
> 
> Hi @DubeySandeep @U8NWXD @StephenYu2018 Please have a look.
> 
> Please note that currently I am facing one of the test failures which I am trying to figure out, pasting the ss for the reference. I have introduced one modal for deleting the user-group and due to that I am getting the "scss file not present" error, as far as I have note we do not really have scss file for modals. Please let me know if you have some info to avoid the error. Failing CI test link - [link](https://github.com/oppia/oppia/actions/runs/10136933232/job/28026571162?pr=20429) Thanks ![Screenshot from 2024-07-29 08-39-36](https://private-user-images.githubusercontent.com/44300735/352907849-7aa1fb60-dde5-4c3a-a197-429d37c643ee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NzgwOTUsIm5iZiI6MTcyMjg3Nzc5NSwicGF0aCI6Ii80NDMwMDczNS8zNTI5MDc4NDktN2FhMWZiNjAtZGRlNS00YzNhLWExOTctNDI5ZDM3YzY0M2VlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDE3MDk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjNzU0NjllZGZkMjU0NWIwMTdkMjgwZGE3MTFmMzQ1MGFiODE1YTQzNDhkYjM0MjFlMWEyN2U2YjQ1YTMzNjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gwKsz4oYOQ43cCUrfqI5bWEu__s7erxZOYyGr8BPmM0)

This issue is resolved!

seanlip commented on core/domain/user_services.py:
```diff
@@ -495,6 +495,96 @@ def get_user_settings_by_auth_id(
+    name and value as list of users associated to the user-group.
+
+    Returns:
+        user_group_models_dict: List[Dict[str, Union[str, List[str]]]]. List of
```
You should be returning domain objects internally, not dicts.

Also I think user_group_models_dict is not the right name because the dict doesn't contain models in it. Differentiate clearly between storage models (which should not be used except at the interface to the datastore) and domain objects (which are plain Python objects that can be used throughout the code).

lkbhitesh07 replied:
Make sense, I have updated this.

seanlip commented on core/domain/user_services.py:
```diff
@@ -495,6 +495,96 @@ def get_user_settings_by_auth_id(
+            if not user in all_users_usernames:
+                raise Exception(
+                    f'The user {user} of user-group {old_user_group_name} '
+                    'does not exists.')
```
exists --> exist

lkbhitesh07 replied:
Updated

seanlip commented on core/domain/user_services.py:
```diff
@@ -495,6 +495,96 @@ def get_user_settings_by_auth_id(
+    user_group_model.delete()
+
+
+def update_user_group(
```
I don't think you're modelling this correctly. I think what you need is an immutable ID for each user group (set upon creation), and a human-readable name which can be changed.

This is because, when you specify a user group in the config for a feature flag, you'll need to do that specification by ID. If you do it by a mutable name, then if that name changes, the feature flags will point to an incorrect reference.

lkbhitesh07 replied:
Got it, I have updated this.

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -58,6 +59,117 @@ def delete(self) -> None:
+        assert self.normalized_payload is not None
+        action = self.normalized_payload.get('action')
+        try:
+            if action == 'update_user_group':
```
I think this isn't the right modelling. Use put() for updating and delete() for deleting, and pass simpler payloads -- you don't need an "action" field here, it's overkill.

lkbhitesh07 replied:
Updated!

seanlip reviewed (changes requested):
Hi @lkbhitesh07, sorry for the late reply. I did a feature review and I think the UI looks quite good! Some notes:

- The title of the section should be "Manage User Groups", not "Update UserGroups". (Note also the space between "User and "Groups".)
- Have the "Save" button be furthest on the right, and the "Cancel" button be on its left (and white in colour, not yellow).
- I think the delete button should always show in "expanded" mode -- it's not intuitive that you need to edit the list of users in order to see the delete button.
- How does the user specify a user group in a feature flag config?
- When the user attempts to delete a user group, will the deletion be prevented if it's currently being used in any feature flag configs?
- What happens if the user renames an existing user group (usergroup3 in your video) to another existing user group (e.g. "usergroup2")? 
- After the name of usergroup1 is changed to usergroup3, is it possible to rename an existing user group to usergroup 1?

Also I had some code comments, PTAL. Thanks!

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

StephenYu2018 commented on core/domain/user_services.py:
```diff
@@ -495,6 +495,96 @@ def get_user_settings_by_auth_id(
+    name and value as list of users associated to the user-group.
+
+    Returns:
+        user_group_models_dict: List[Dict[str, Union[str, List[str]]]]. List of
```
Return value shouldn't be named in the function docstring.

lkbhitesh07 replied:
Got it, I have updated this.

StephenYu2018 reviewed (changes requested):
One concern.

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Added the domain layer changes

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Fixed backend test and updated the UI

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 commented:
Hi @seanlip, thanks for the review. I have made the changes, please have a look.


>The title of the section should be "Manage User Groups", not "Update UserGroups". (Note also the space between "User and "Groups".)

Done!


>Have the "Save" button be furthest on the right, and the "Cancel" button be on its left (and white in colour, not yellow).

Done!

>I think the delete button should always show in "expanded" mode -- it's not intuitive that you need to edit the list of users in order to see the delete button.

Done!

>How does the user specify a user group in a feature flag config?

I will be raising a follow up PR to handle this.

>When the user attempts to delete a user group, will the deletion be prevented if it's currently being used in any feature flag configs?

Yes, we will be handling this scenario in my follow up PR.

>What happens if the user renames an existing user group (usergroup3 in your video) to another existing user group (e.g. "usergroup2")?

We show an error message. I have updated the video also here is the ss
![Screenshot from 2024-09-25 07-54-23](https://github.com/user-attachments/assets/481af1ee-7343-45b3-adf6-3e9d5729503e)

>After the name of usergroup1 is changed to usergroup3, is it possible to rename an existing user group to usergroup 1?

Yes, we can do that. I have updated the video.

Thanks!

lkbhitesh07 committed:
fixed tests

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

seanlip commented on core/domain/user_domain.py:
```diff
@@ -526,6 +526,94 @@ def mark_lesson_info_modal_viewed(self) -> None:
+        """
+        self.user_group_id = user_group_id
+        self.name = name
+        self.users = users
```
"users" is unclear -- if you're just talking about the strings then use "user_ids". Otherwise it sounds like a list of User objects.

There's also a point to be careful about here -- the datastore treats user IDs as the immutable ID for a user (and this is not guaranteed with user names or emails), but the release coordinator is entering usernames. So in the data model I think you should be storing user IDs, and we should be passing usernames out to the controller.

But there's a question about what this domain object should be storing. I think probably usernames is better and you can convert to/from user IDs at the time of datastore fetch/save, and nowhere else...but what do you think? (All this is also another reason why it is pretty important to disambiguate this field name!)

lkbhitesh07 replied:
I see, I agree with you. I have made the changes and have opted for usernames in the domain layer.

seanlip commented on core/domain/user_domain.py:
```diff
@@ -526,6 +526,94 @@ def mark_lesson_info_modal_viewed(self) -> None:
+            raise utils.ValidationError(
+                'Expected name to be a string, received %s.' % self.name)
+
+        if not isinstance(self.users, list):
```
You also need to validate that each element of this is a string.

lkbhitesh07 replied:
Done

seanlip commented on core/domain/user_domain.py:
```diff
@@ -526,6 +526,94 @@ def mark_lesson_info_modal_viewed(self) -> None:
+        self.name = updated_name
+        self.validate()
+
+    def update_users(self, updated_users: List[str]) -> None:
```
update_user_ids (and similar changes throughout)

lkbhitesh07 replied:
Updated

seanlip commented on core/domain/user_services.py:
```diff
@@ -499,6 +499,139 @@ def get_user_settings_by_auth_id(
         return None
 
 
+def get_all_user_group() -> List[user_domain.UserGroup]:
```
get_all_user_groups

lkbhitesh07 replied:
Done

seanlip commented on core/domain/user_services.py:
```diff
@@ -499,6 +499,139 @@ def get_user_settings_by_auth_id(
+        user_models.UserGroupModel.get_all())
+    user_groups_list = []
+    for user_group_model in user_group_models:
+        user_group_dict: user_domain.UserGroupDict = {
```
Try to avoid using dicts in parts of the code that are not application boundaries. Just create the UserGroup domain object directly using the constructor.

lkbhitesh07 replied:
Updated

seanlip commented on core/domain/user_services.py:
```diff
@@ -499,6 +499,139 @@ def get_user_settings_by_auth_id(
+    Raises:
+        Exception. The user inside user group does not exist.
+    """
+    all_users_usernames = get_all_users_usernames()
```
Woah, what is this? Don't do this, there is no need to retrieve several thousand usernames when you just need to get a few.

Depending on the data modelling question above, if you are storing usernames in the domain object, then I think you'll need a new OR query in the storage layer that queries for whether the username = username1 OR username2 OR ..., so that you can get the results in one go. In the case where some usernames are missing, then you can formulate an error message by fetching the usernames corresponding to the userIDs in a separate query, and subtracting those from the original list of usernames to get the invalid ones

lkbhitesh07 replied:
I see, This make sense. I have made the changes, I am using the OR query for this section.

seanlip commented on core/domain/user_services.py:
```diff
@@ -499,6 +499,139 @@ def get_user_settings_by_auth_id(
+    if len(user_group_users) != len(set(user_group_users)):
+        raise Exception(
+            f'Users list of user-group {user_group_name} contains '
+            'duplicates.'
```
State, in the error message, what the duplicates are. (You can perhaps do that by subtracting the set from the original list and then applying set() to the result to make the elements unique, though there might be simpler ways.)

lkbhitesh07 replied:
Got it, updated.

seanlip commented on core/domain/user_services.py:
```diff
@@ -499,6 +499,139 @@ def get_user_settings_by_auth_id(
+            f'Users list of user-group {user_group_name} contains '
+            'duplicates.'
+        )
+    user_group_dict: user_domain.UserGroupDict = {
```
Same note, avoid creating dicts at non-application boundaries.

lkbhitesh07 replied:
Updated

seanlip commented on core/domain/user_services.py:
```diff
@@ -1390,6 +1523,19 @@ def set_username(user_id: str, new_username: str) -> None:
     save_user_settings(user_settings)
 
 
+def get_all_users_usernames() -> List[str]:
```
We shouldn't be doing this. See above.

lkbhitesh07 replied:
Updated

seanlip replied:
Er, we still should not be doing this. There should be no function here that does a get_all() on tens to hundreds of thousands of users and iterates over them. That takes too long.

lkbhitesh07 replied:
Oops! Sorry, I thought that I removed it. Updated now.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -2380,14 +2380,13 @@ class UserGroupModel(base_models.BaseModel):
     ID_IS_USED_AS_TAKEOUT_KEY: Literal[True] = True
 
     # The user ids of the users.
+    name = datastore_services.StringProperty(required=True, indexed=True)
```
The comment above is wrong, it does not describe this line,

The name below should also be clarified ("users" is vague).

lkbhitesh07 replied:
Updated

seanlip reviewed (changes requested):
Thanks @lkbhitesh07, sorry for the late reply. UI looks pretty good, some comments:

- Please have the success toast be a bit more specific -- at least say which of the user groups was updated. (The idea that, if you do a small change, you end up updating all user groups at once, is a bit scary.)
- Can you restrict the names of the user groups to alphanumeric characters and spaces? Please show the error message when that's not done. Include backend validation too (as a stopgap).

I also left inline comments in the code about the implementation, PTAL. Thanks!

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

StephenYu2018 reviewed (approved):
LGTM!

lkbhitesh07 committed:
Done suggested changes: Removed the fetching of all users

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 commented:
> Thanks @lkbhitesh07, sorry for the late reply. UI looks pretty good, some comments:
> 
>     * Please have the success toast be a bit more specific -- at least say which of the user groups was updated. (The idea that, if you do a small change, you end up updating all user groups at once, is a bit scary.)
> 
>     * Can you restrict the names of the user groups to alphanumeric characters and spaces? Please show the error message when that's not done. Include backend validation too (as a stopgap).
> 
> 
> I also left inline comments in the code about the implementation, PTAL. Thanks!

Thanks for the review @seanlip, I have made the changes, below is the ss for the changes:
![Screenshot from 2024-10-06 18-02-40](https://github.com/user-attachments/assets/f312dc2e-0550-41af-a10b-db6a9b0de042)

![Screenshot from 2024-10-06 18-03-40](https://github.com/user-attachments/assets/c1a30104-22cd-4cfa-b83d-b0cb08bb70e2)

lkbhitesh07 committed:
small nit

lkbhitesh07 committed:
backend coverage

lkbhitesh07 committed:
fixed tests

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -58,6 +59,148 @@ def delete(self) -> None:
+    """
+
+    user_group_name: Optional[str]
+    user_group_user_usernames: Optional[List[str]]
```
I think you can just call this "member_names" (the current name is a bit clumsy). Arrange the properties in the order: user_group_id, user_group_name, member_names.

lkbhitesh07 replied:
I see, I have updated it to member_usernames.

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -58,6 +59,148 @@ def delete(self) -> None:
+                'default_value': None
+            }
+        },
+        'POST': {
```
Normally we arrange in the order GET/POST/PUT/DELETE, I think. Ditto for the handlers below.

lkbhitesh07 replied:
Updated

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -58,6 +59,148 @@ def delete(self) -> None:
+        assert self.user_id is not None
+        assert self.normalized_payload is not None
+        try:
+            user_group_name = (
```
Why do you have a default_value of None in the handler schema if we don't accept None values for the update? Should those default values be dropped?

lkbhitesh07 replied:
Right, I have updated this.

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -58,6 +59,148 @@ def delete(self) -> None:
+            user_services.update_user_group(
+                user_group_id, user_group_name, user_group_user_usernames)
+            self.render_json(self.values)
+        except Exception as e:
```
Why do we need the try/except here and in post() when we don't do this for other handlers (is there something we want to achieve here besides the default logging behaviour)?

lkbhitesh07 replied:
Make sense, I have removed it.

seanlip commented on core/domain/feature_flag_services.py:
```diff
@@ -245,7 +245,7 @@ def is_feature_flag_enabled(
+                user_models.UserGroupModel.user_ids == user_id
             ).fetch())
 
         user_group_models_ids: List[str] = []
```
Make this a set so that the lookup in line 256 is faster. I think you can also use a list comprehension to populate the set in a single line.

lkbhitesh07 replied:
Updated.

seanlip commented on core/domain/user_domain.py:
```diff
@@ -526,6 +526,114 @@ def mark_lesson_info_modal_viewed(self) -> None:
+        self,
+        user_group_id: str,
+        name: str,
+        user_usernames: List[str]
```
member_usernames

lkbhitesh07 replied:
Updated

seanlip commented on core/domain/user_domain.py:
```diff
@@ -526,6 +526,114 @@ def mark_lesson_info_modal_viewed(self) -> None:
+            user_usernames: List[str]. The list of user usernames
+                attached to user group.
+        """
+        self.user_group_id = user_group_id
```
I'm getting a bit confused now ... in the controiller layer you say user_group_name and not name. Would it be possible to keep things consistent up and down the stack where feasible?

(In general I think "name" is probably fine since we'd typically write `user_group.name`.)

lkbhitesh07 replied:
Oh, right. I have updated everywhere.

seanlip commented on core/domain/user_services.py:
```diff
@@ -499,6 +501,161 @@ def get_user_settings_by_auth_id(
+    user_groups_list = []
+    for user_group_model in user_group_models:
+        user_usernames: List[str] = []
+        for user_setting in get_users_settings(user_group_model.user_ids):
```
user_settings

lkbhitesh07 replied:
Updated

seanlip commented on core/domain/user_services.py:
```diff
@@ -499,6 +501,161 @@ def get_user_settings_by_auth_id(
+    for user_group_model in user_group_models:
+        user_usernames: List[str] = []
+        for user_setting in get_users_settings(user_group_model.user_ids):
+            assert user_setting is not None
```
Do you need some sort of custom error message to say which username is invalid, so that whoever is debugging knows what exactly needs to be investigated? I think with assert statements in Python you can add an optional error message.

Ditto below.

lkbhitesh07 replied:
I see, I have added the custom message, we will be able to see the user_id of the user whose username is not valid.

seanlip commented on core/domain/user_services.py:
```diff
@@ -499,6 +501,161 @@ def get_user_settings_by_auth_id(
+        user_models.UserSettingsModel.username == username
+        for username in user_group_user_usernames
+    ]
+    existing_users: Sequence[user_models.UserSettingsModel] = (
```
existing_users_settings

lkbhitesh07 replied:
Updated

seanlip commented on core/domain/user_services.py:
```diff
@@ -499,6 +501,161 @@ def get_user_settings_by_auth_id(
+        ).fetch()
+    )
+    existing_users_usernames = [
+        user_setting.username for user_setting in existing_users
```
for user_settings in existing_users_settings

lkbhitesh07 replied:
Updated

seanlip commented on core/domain/user_services.py:
```diff
@@ -499,6 +501,161 @@ def get_user_settings_by_auth_id(
+            which validation needs to be done.
+
+    Raises:
+        Exception. The user inside user group does not exist.
```
Your function has two types of exceptions, but this only describes one.

lkbhitesh07 replied:
Updated

seanlip commented on core/feconf.py:
```diff
@@ -1047,6 +1047,7 @@ def get_empty_ratings() -> Dict[str, int]:
     '/updatetranslationsuggestionhandler')
 UPDATE_QUESTION_SUGGESTION_URL_PREFIX = (
     '/updatequestionsuggestionhandler')
+USER_GROUPS_HANDLER_URL = '/user_groups'
```
Is this a handler URL? If so should it end with handler? How do we name other such URLs, generally?

lkbhitesh07 replied:
Right, I have updated this. I have added "_handler" at the end, following other patterns.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -2380,20 +2380,18 @@ class UserGroupModel(base_models.BaseModel):
 
-    # The user ids of the users.
-    users = datastore_services.StringProperty(repeated=True)
+    name = datastore_services.StringProperty(required=True, indexed=True)
```
There should be a comment above each of these lines with the correct docstring.

You might also want to consider calling user_ids below member_ids instead, to match member_names. This is optional though.

lkbhitesh07 replied:
Updated

seanlip commented on core/templates/pages/release-coordinator-page/release-coordinator-page.component.ts:
```diff
@@ -46,16 +64,34 @@ export class ReleaseCoordinatorPageComponent implements OnInit {
+  loadingMessage: string = '';
+  newUserGroupName: string = '';
+  separatorKeysCodes: number[] = [ENTER];
+  userGroupsBackup!: Map<string, UserGroup>;
```
Could you add a comment explaining what this variable stores and describing the lifecycle for backups?

lkbhitesh07 replied:
Updated

seanlip commented on core/templates/pages/release-coordinator-page/release-coordinator-page.component.html:
```diff
@@ -73,9 +76,159 @@ <h3>Enable Promo Bar</h3>
+        <button class="btn oppia-add-new-user-group-btn"
+                (click)="addUserGroup()"
+                [disabled]="userGroupInEditMode">
+          + Add New UserGroup
```
Add New User Group

lkbhitesh07 replied:
Updated

seanlip commented on core/templates/pages/release-coordinator-page/release-coordinator-page.component.html:
```diff
@@ -73,9 +76,159 @@ <h3>Enable Promo Bar</h3>
+        </div>
+      </div>
+      <div class="oppia-add-user-group">
+        <input placeholder="New UserGroup Name"
```
New user group name

(For user-facing text, in general, "user group" is 2 words.)

lkbhitesh07 replied:
Updated

seanlip commented on core/templates/pages/release-coordinator-page/release-coordinator-page.component.ts:
```diff
@@ -105,6 +141,273 @@ export class ReleaseCoordinatorPageComponent implements OnInit {
+      )
+      .then(
+        () => {
+          this.statusMessage = `UserGroup ${userGroup.userGroupName} successfully updated.`;
```
UserGroup --> User group . Please fix similarly elsewhere for user-facing messages.

Also, put quotes around the user group name anytime you interpolate it within a sentence.

lkbhitesh07 replied:
Updated

seanlip reviewed (changes requested):
Thanks @lkbhitesh07. Took another pass, PTAL.

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 committed:
Done suggested changes

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 commented:
Hi @seanlip, I have addressed all the comments.
Thanks

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -142,63 +138,39 @@ def get(self) -> None:
+        """Performs series of action based on action parameter for user
+        groups.
+        """
         assert self.user_id is not None
```
Does this not already happen in the can_access_release_coordinator_page decorator?

lkbhitesh07 replied:
Ah, right. I have updated it.

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -142,63 +138,39 @@ def get(self) -> None:
     @acl_decorators.can_access_release_coordinator_page
     def put(self) -> None:
         """Updates the specified user group."""
         assert self.user_id is not None
```
Ditto.

lkbhitesh07 replied:
Updated

seanlip commented on core/domain/user_services.py:
```diff
@@ -511,85 +511,92 @@ def get_all_user_groups() -> List[user_domain.UserGroup]:
+        member_usernames: List[str] = []
+        user_ids = user_group_model.user_ids
+        user_settings_list = get_users_settings(user_ids, strict=True)
+        for user_id, user_settings in zip(user_ids, user_settings_list):
```
Thinking about this a bit more -- let's say we create a user group with user X in it, then user X deletes their account (in the self-service way through the preferences page). How is this handled when we next reload the user groups page? Is there a backend test that covers this scenario?

lkbhitesh07 replied:
Right, make sense. This part is being handled in the wipeout_services.py, when we delete the user it will be deleted from all the UserGroupModel objects.
The test have been added in the wipeout_services_test.py file.

seanlip commented on core/templates/pages/release-coordinator-page/release-coordinator-page.component.ts:
```diff
@@ -191,7 +203,7 @@ export class ReleaseCoordinatorPageComponent implements OnInit {
         this.backendApiService.deleteUserGroupAsync(userGroupId).then(
           () => {
-            this.statusMessage = 'UserGroups successfully deleted.';
+            this.statusMessage = 'User Group successfully deleted.';
```
User group XXX successfully deleted.

(Note that "group" is not capitalized -- see my [previous comment](https://github.com/oppia/oppia/pull/20429#discussion_r1789577446). Please fix elsewhere too.)

lkbhitesh07 replied:
Right, updated.

seanlip commented on core/templates/pages/release-coordinator-page/release-coordinator-page.component.html:
```diff
@@ -173,15 +173,15 @@ <h3>Manage User Groups</h3>
       </div>
       <div class="oppia-add-user-group">
-        <input placeholder="New UserGroup Name"
+        <input placeholder="New User Group Name"
```
New user group name

lkbhitesh07 replied:
Updated.

seanlip reviewed (changes requested):
Thanks @lkbhitesh07 -- took a pass, just a few things left. PTAL!

lkbhitesh07 committed:
user also gets deleted from UserGroupModel when it is deleted

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

lkbhitesh07 commented:
Hey @seanlip, thanks for the review. I have addressed all the comments.

lkbhitesh07 committed:
Fixed coverage

seanlip commented on core/domain/wipeout_service_test.py:
```diff
@@ -1225,6 +1225,23 @@ def test_one_config_property_is_pseudonymized(self) -> None:
+        wipeout_service.delete_user(
+            wipeout_service.get_pending_deletion_request(self.user_1_id))
+
+        updated_user_group = user_models.UserGroupModel.get_by_id('group_1')
```
updated_user_group_model (please differentiate between models and domain objects)

lkbhitesh07 replied:
Done

seanlip commented on core/domain/wipeout_service_test.py:
```diff
@@ -1225,6 +1225,23 @@ def test_one_config_property_is_pseudonymized(self) -> None:
+        )
+        user_group_model.put()
+
+        user_group = user_models.UserGroupModel.get_by_id('group_1')
```
existing_user_group_model

lkbhitesh07 replied:
Done

seanlip reviewed (approved):
Thanks @lkbhitesh07! Two trivial comments otherwise LGTM. Feel free to submit once you address them.

lkbhitesh07 committed:
Some trivial changes

lkbhitesh07 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-user-group

seanlip commented:
This failed merge queue twice for reasons unrelated to the PR, so I'm directly merging it.

I've also added comments to issues #20958 and #21145 to report the failing tests.
